### DISCUSSION TITLE
Idea: Have Copilot automatically identify places it thinks you may have an error

### DISCUSSION THREAD

**Discussion by andenacitelli** [community/community#11459] [Copilot Conversations]
**Background:** With many languages, especially untyped ones like JavaScript or Python, you usually don't figure out whether you have an error in your code until runtime, and even then you may only get to that code under very specific circumstances. While things like TypeScript exist and can do a great job of identifying these, they involve some additional setup that can at times be annoying to configure. 

**Feature Proposal:** Analyze code and identify/suggest places your code likely has an error. Copilot may particularly excel at this because it's able to generalize past just typechecking. 

**Implementation:** A naive approach I can imagine would be to have Copilot check to make sure every line of a file is the same as what Copilot's suggestion for it would be if that line were deleted. Sure, it doesn't sound very performant at first consideration, but I don't know enough about Copilot behind the scenes to say for sure. It can potentially be optimized by having it analyze entire functions, blocks of code, or files, at once, or by doing some pre-analysis before running it through the model in order to trim the lines to actually test in the first place. 

**Potential Related Feature:** Have Copilot identify places you could do something in a more performant way. Maybe you're doing an expensive linear search on a sorted array or something and a binary search can instead be utilized. 

My apologies if this has already been suggested. Love the work you guys have been doing - Copilot already saves me a bunch of time and mental energy and I can't wait to see it become even better!

**Comment by KTibow** [comment]
This sounds very interesting. However, there's a couple of things that I think might not work:

> A naive approach I can imagine would be to have Copilot check to make sure every line of a file is the same as what Copilot's suggestion for it would be if that line were deleted.

Copilot won't replicate code, and is nowhere near perfect. Of course, as you stated, this is a naive approach.

> but I don't know enough about Copilot behind the scenes to say for sure.

It's simpler than you think. You give Copilot some code, and it predicts what goes next. You can see what code exactly is being fed to Copilot under Terminal > Output > GitHub Copilot.

In the end, I feel like this might have to be a seperate AI than the original Codex model - given that (AFAIK) it was trained on the latest revision of popular open source code, which probably doesn't have that many bugs.

**Reply by andenacitelli** [reply]
> Copilot won't replicate code, and is nowhere near perfect.

I apologize if this is essentially a restatement of the question, but in your words, "You give Copilot some code, and it predicts what goes next.". Why couldn't this "code" just be the prior lines of the file to the one you're interested in checking? I unfortunately don't think I fully understand what you mean by it not being able to "replicate code" and why it prevents this. 
 
> Copilot [...] is nowhere near perfect.

I understand, by virtue of being a model trained on user content, it's fundamentally far from perfect. However, in my experience, it is usually at least _close_, and sometimes exactly what you need, especially in very common or "boilerplate" scenarios. My hope is that the code it would suggest is close enough to "correct" to at least give you a "hey, this code looks a little funky" with reasonable confidence. You'd want a high true positive rate and a low false negative rate to be useful, else you'd probably start to ignore its suggestions, as they tended to not be worth the time to look at. 

> You can see what code exactly is being fed to Copilot under Terminal > Output > GitHub Copilot.

This only seems to show me the API endpoints, rather than the actual payloads. Is something like Wireshark necessary for me to view these, or is there a built-in method you're aware of that I can use to get more detail? Here's the terminal output I get trying out some simple predictions: 
![image](https://user-images.githubusercontent.com/30580900/154404790-0265cd89-c7a1-467a-b0bc-00313be661c3.png)

> In the end, I feel like this might have to be a seperate AI than the original Codex model - given that (AFAIK) it was trained on the latest revision of popular open source code, which probably doesn't have that many bugs.

Yeah, seems like there's at least two potential approaches here. You either get a model trained on (reasonably) bug-free code and simply see if its suggestion _doesn't_ match your code, or you get a model (or models) trained on buggy code and see if it's suggestion _does_ match your code. I personally think the first instance feels more generalizable and more easily trainable, which feels more in the spirit of Copilot. The second almost essentially just feels like a glorified linter or "spellcheck" with pre-programmed errors it would point out. However, I'm no machine learning engineer, so would be entirely open to input on your or others' part.

**Reply by KTibow** [reply]
> I unfortunately don't think I fully understand what you mean by it not being able to "replicate code" and why it prevents this.

Yes, when there's a universal way to solve a problem, it'll probably be correct, but there are often thousands of reasonable ways to complete a line. And as you probably have experienced, you can't just constantly accept Copilot's suggestions.

> This only seems to show me the API endpoints,

Wait, they stopped doing that? :facepalm: I recommended that to so many people.
